core/executor/src/allocator.rs

```diff
@@ -0,0 +1,471 @@
+		let current_size = current_size.0 as u32;
+		let used_size = mem.used_size().0 as u32;
+		let heap_size = current_size - used_size;
+
+		let mut ptr_offset = used_size;
+		let padding = ptr_offset % ALIGNMENT;
+		if padding != 0 {
+			ptr_offset += ALIGNMENT - padding;
+		}
+
+		FreeingBumpHeapAllocator {
+			bumper: 0,
+			heads: [0; N],
+			heap: mem,
+			max_heap_size: heap_size,
+			ptr_offset: ptr_offset,
+			total_size: 0,
+		}
+	}
+
+	/// Gets requested number of bytes to allocate and returns a pointer.
+	/// The maximum size which can be allocated at once is 16 MiB.
+	pub fn allocate(&mut self, size: u32) -> u32 {
+		if size > MAX_POSSIBLE_ALLOCATION {
+			return 0;
+		}
+
+		let size = size.max(8);
+		let item_size = size.next_power_of_two();
+		if item_size + 8 + self.total_size > self.max_heap_size {
+			return 0;
+		}
+
+		let list_index = (item_size.trailing_zeros() - 3) as usize;
+		let ptr: u32 = if self.heads[list_index] != 0 {
+			// Something from the free list
+			let item = self.heads[list_index];
+			self.heads[list_index] = FreeingBumpHeapAllocator::le_bytes_to_u32(&mut self.get_heap(item, 4));
+			item + 8
+		} else {
+			// Nothing to be freed. Bump.
+			self.bump(item_size + 8) + 8
+		};
+
+		for i in 1..8 { self.set_heap(ptr - i, 255); }
+
+		self.set_heap(ptr - 8, list_index as u8);
+
+		self.total_size = self.total_size + item_size + 8;
+		trace!(target: "wasm-heap", "Heap size is {} bytes after allocation", self.total_size);
+
+		self.ptr_offset + ptr
+	}
+
+	/// Deallocates the space which was allocated for a pointer.
+	pub fn deallocate(&mut self, ptr: u32) {
+		let ptr = ptr - self.ptr_offset;
+
+		let list_index = self.get_heap_byte(ptr - 8) as usize;
+		for i in 1..8 { assert!(self.get_heap_byte(ptr - i) == 255); }
+		let tail = self.heads[list_index];
+		self.heads[list_index] = ptr - 8;
+
+		let slice = &mut self.get_heap(ptr - 8, 4);
+		FreeingBumpHeapAllocator::write_u32_into_le_bytes(tail, slice);
+		self.set_heap_bytes(ptr - 8, slice);
+
+		let item_size = FreeingBumpHeapAllocator::get_item_size_from_index(list_index);
+		self.total_size = self.total_size.checked_sub(item_size as u32 + 8).unwrap_or(0);
+		trace!(target: "wasm-heap", "Heap size is {} bytes after deallocation", self.total_size);
+	}
+
+	fn bump(&mut self, n: u32) -> u32 {
+		let res = self.bumper;
+		self.bumper += n;
+		res
+	}
+
+	fn le_bytes_to_u32(slice: &mut [u8]) -> u32 {
+		let bytes = [slice[0], slice[1], slice[2], slice[3]];
+		unsafe { std::mem::transmute::<[u8; 4], u32>(bytes) }.to_le()
+	}
+
+	fn write_u32_into_le_bytes(bytes: u32, slice: &mut [u8]) {
+		let bytes: [u8; 4] = unsafe { std::mem::transmute::<u32, [u8; 4]>(bytes.to_le()) };
+		for i in 0..4 { slice[i] = bytes[i]; }
+	}
+
+	fn get_item_size_from_index(index: usize) -> usize {
+		// we shift 1 by three places, since the first possible item size is 8
+		1 << 3 << index
+	}
+
+	fn get_heap(&mut self, ptr: u32, size: usize) -> Vec<u8> {
```

Comment:
since we only use it to interpret either 1 or 4 byte values, we can just use `get_into()` to avoid dynamic allocation.